core.ts

```diff
@@ -980,6 +988,56 @@ abstract class PinejsClientCoreTemplate<
 		return this.request(params, { method: 'DELETE' });
 	}
 
+	upsert(params: PinejsClientCoreFactory.UpsertParams) {
+		const { id, body, ...restParams } = params;
+
+		if (!isObject(id)) {
+			throw new Error('The id property must be an object');
+		}
+
+		const naturalKeyProps = Object.keys(params.id);
+		if (naturalKeyProps.length === 0) {
+			throw new Error(
+				'The id property must be an object with the natural key of the model',
+			);
+		}
+
+		if ('options' in params) {
+			throw new Error(`'options' are not supported in upserts`);
+		}
+
+		if (body == null) {
+			throw new Error('The body property is missing');
+		}
+
+		const postParams = {
+			...restParams,
+			body: {
+				...body,
+				...id,
+			},
+		};
+		return this.post(postParams).catch(err => {
```

Comment:
I'd prefer
```suggestion
		return this.post(postParams).then(undefined, err => {
```
to avoid the need for the major change and drop this to just a minor change, that way this change (and any future ones) can be much more easily propagated to all of its dependents